oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/spi/editor/FulltextIndexEditor.java

```diff
@@ -144,8 +145,11 @@ public void leave(NodeState before, NodeState after)
             }
         }
 
-        for (Aggregate.Matcher m : matcherState.affectedMatchers) {
-            m.markRootDirty();
+        for (int i = 0; i < matcherState.matched.size(); i++) {
+            if (matcherState.affectedMatchers.get(i)) {
+                Aggregate.Matcher m = matcherState.matched.get(i);
+                m.markRootDirty();
+            }
```

Comment:
Something like this:
```suggestion
        BitSet bitSet = matcherState.affectedMatchers;
        for (int i = bitSet.nextSetBit(0); i != -1; i = bitSet.nextSetBit(i + 1)) {
            Aggregate.Matcher m = matcherState.matched.get(i);
            m.markRootDirty();
        }
```

Reply:
Good idea, I changed the code to iterate over the BitSet instead over all the matchers.